MobileCenter/MobileCenter/Internals/Channel/MSLogManagerDefault.m

```diff
@@ -223,18 +239,134 @@ - (void)setEnabled:(BOOL)isEnabled andDeleteDataOnDisabled:(BOOL)deleteData forG
+              }];
+            }
+            
+            // Init block to handle all the channels "done flushing" notifications.
+            MSDoneFlushingCompletionBlock completion = ^(){
+              typeof(self) toughSelf = weakSelf;
+                if (toughSelf) {
+                  @synchronized(toughSelf.backgroundTaskLockToken) {
+                    
+                    // Background task is still going on.
+                    if (toughSelf.backgroundTaskIdentifier != UIBackgroundTaskInvalid) {
+                      toughSelf.flushedChannelsCount ++;
+                      
+                      // All channels have finished flushing.
+                      if (toughSelf.channels.count == toughSelf.flushedChannelsCount){
+                        
+                        // Disable sender so that network changes in background won't affect his state.
+                        [toughSelf.sender setEnabled:NO andDeleteDataOnDisabled:NO];
+                        [toughSelf endBackgroundActivity];
+                      }
+                    }
+                  }
+                }
+            };
+            
+            // There is now extended time to flush the last few logs from the channels.
+            for (NSString *groupId in strongSelf.channels) {
+              [strongSelf.channels[groupId] notifyWhenDoneFlushingWithCompletion:completion];
+            }
+          }
+        };
+        self.appDidEnterBackgroundObserver =
+        [MS_NOTIFICATION_CENTER addObserverForName:UIApplicationDidEnterBackgroundNotification
+                                            object:nil
+                                             queue:NSOperationQueue.mainQueue
+                                        usingBlock:notificationBlock];
+      }
+      self.appWillEnterForegroundObserver =
+      [MS_NOTIFICATION_CENTER addObserverForName:UIApplicationWillEnterForegroundNotification
+                                          object:nil
+                                           queue:NSOperationQueue.mainQueue
+                                      usingBlock:^(NSNotification __unused *note) {
+                                        typeof(self) strongSelf = weakSelf;
+                                        if (strongSelf) {
+                                          @synchronized(strongSelf.backgroundTaskLockToken) {
+
+                                            // In foreground now, cancel any pending background task.
+                                            [strongSelf endBackgroundActivity];
+                                            
+                                            // Re-enable sender.
+                                            [strongSelf.sender setEnabled:YES andDeleteDataOnDisabled:NO];
+                                            
+                                            /**
+                                             * TODO: Needs some refactoring, sender is coupled to too many objects.
```

Comment:
Yes. It'd be great if we could decouple this soon. Maybe time to draw & implement a state machine.

Reply:
We can try that, although I have some ideas already.